src/tactic/suggest.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2019 Lucas Allen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lucas Allen and Scott Morrison
+-/
+
+import tactic.library_search
+import data.mllist
+
+/-!
+`suggest` is an extension of `library_search` which lists some applicable lemmas and theorems.
+-/
+
+namespace tactic
+
+open tactic.library_search
+
+/--
+Assuming that a goal `g` has been (partially) solved in the tactic_state `l`,
+this function prepares a string of the form `exact ...` or `refine ...` (possibly with underscores)
+that will reproduce that result.
+-/
+meta def message (g : expr) (l : tactic_state) : tactic string :=
+do s ← read,
+   write l,
+   r ← tactic_statement g,
+   write s,
+   return r
+
+/--
+Runs through the list of tactic_states in which the goal `g` has been (partially) solved,
+and generates an `exact ...` or `refine ...` message for each one.
+-/
+meta def messages (g : expr) (L : list tactic_state) : tactic (list string) :=
+L.mmap (message g)
+
+declare_trace silence_suggest -- Turn off `exact/refine ...` trace messages
+declare_trace suggest         -- Trace a list of all relevant lemmas
+
+/--
+The main `suggest` tactic, which is very similar to the main `library_search` function.
+It returns a list of pairs `(state, message)` consisting of
+* `state`, a tactic state resulting from the successful application of a declaration from the library, and
+* `message`, a string of the form `refine ...` or `exact ...` which will reproduce that tactic state.
+-/
+meta def suggest (limit : option ℕ := none) (discharger : tactic unit := done) : tactic (list (tactic_state × string)) :=
+do [g] ← get_goals | fail "`suggest` should be called with exactly one goal",
```

Comment:
Is this restriction really necessary? Rather than require one goal, you could just focus on the first goal. (Same goes for `library_search`.) Most of the time we hope users will have done that already. But these tactics are very useful for beginners who might not be writing good structured proofs yet.

Reply:
I fixed this in the big refactor that built `library_search` on top of `suggest_core`. Both tactics are now happy with multiple goals.